docs: finish documenting migration steps for v17 (#1800)
commercetools/merchant-center-application-kit: website/src/releases/2020-10/release-v17.mdx

---
date: 2020-10-14
title: Custom Applications v17
description: |
  The Application Kit packages have been released as a new major version v17.\
  In this release note we go through the important migration steps.
type: announcement
topics:
  - Breaking Changes
  - Dependencies
---

The Application Kit packages have been released as a new **major version** `v17`. One of the most important changes in this release is about migrating to the new Apollo Client `v3`.

Follow the steps below to migrate your Custom Application to the new versions.

<!--more-->

# Migrating to Apollo Client v3

Migrating to the new Apollo Client `v3` requires some important migration steps.

## Migrating Apollo imports

Apollo ships now with a single package `@apollo/client` instead of multiple ones.

> For Custom Applications this means that the peer dependencies `apollo-client` and `react-apollo` are now replaced with the new peer dependency of `@apollo/client`.

Reviewer: @emmenko Would nice to include code changes, just like you added in other sections of this release note.
```diff
-import { Query } from 'react-apollo';
+import { Query } from '@apollo/client/react/components';
```

Author: There is a link below to the official apollo docs, where it's properly documented how you change the imports.

I don't think we need to duplicate those examples here as well.

This is all you need to do to migrate to the latest version of Apollo, including of course [updating the Apollo imports](https://www.apollographql.com/docs/react/migrating/apollo-client-3-migration/#updating-imports).

## About the new Apollo Cache

With Apollo `v3`, the Apollo in-memory cache got some improvements as well, resulting in some breaking changes. Please make sure to [read this document](https://www.apollographql.com/docs/react/migrating/apollo-client-3-migration/#cache-improvements) to understand the changes.

Depending on the GraphQL queries and mutations used by your Custom Application, it might be that the caching behavior of those queries needs to be adjusted. For example if objects do not have an `id` field as identifier, or if Apollo cannot automatically merge two objects that have different fields shape, and so on.

> If that is the case, you need to configure the Apollo cache according to your needs, as only you and your Custom Application can decide what is the best caching strategy.

Reviewer 2: ```suggestion
If that is the case, you need to configure the Apollo cache according to your needs, as only you know what the best caching strategy for your Custom Application is.
```

</Info>

## Enforcing a valid context object

The recommended and preferred way of passing request context is to use the `context` object.
To improve the TypeScript support and auto-completion for the **request context** options, the `@commercetools-frontend/application-shell` package now exports new React hooks. Note that autocompletion is possible, even if you are not using TypeScript.

- `useMcQuery`
- `useMcLazyQuery`
- `useMcMutation`

These hooks are thin wrappers around the original Apollo hooks and have the same API, with a minor difference. Namely, the `context` object is properly typed to conform with the Merchant Center **request context** instead of `any`.

```diff
-import { useQuery } from '@apollo/client/react';
+import { useMcQuery } from '@commercetools-frontend/application-shell';

-const { loading, data, error } = useQuery(MyQuery, {
+const { loading, data, error } = useMcQuery(MyQuery, {
  context: {
    target: GRAPHQL_TARGETS.COMMERCETOOLS_PLATFORM,
  },
});
```

# Changes about the `mc-scripts` CLI

## Removing deprecated options from `compile-html`

Previously deprecated CLI options such as `--env`, `--csp`, `--headers`, have been removed.

> Additionally, the CLI flag `--use-local-assets` has been removed as well. The default behavior of `mc-scripts compile-html` now is to compile the assets locally, which is the only reasonable thing to do.

Reviewer 2: ```suggestion
Additionally, the CLI flag `--use-local-assets` has been removed as well. As such the default behavior of `mc-scripts compile-html` changed to compile the assets locally.
```

When running the `mc-scripts compile-html` command, the `index.html` is compiled for production usage and it lives in the `public` folder, together with the other static assets. This is all you need to deploy your application.
You can decide to [deploy the Custom Application statically to one of the popular cloud providers](/deployment/compiling-a-custom-application#deployment), or serve the files on your own using a static server.

For example, to run locally the Custom Application using the production bundles:

```console
NODE_ENV=production MC_APP_ENV=development dotenv -- \
  mc-scripts compile-html \
  --transformer @commercetools-frontend/mc-dev-authentication/transformer-local.js

mc-scripts serve
```

## Removing the `mc-http-server` package

> The `@commercetools-frontend/mc-http-server` package has been deprecated and won't be published anymore.

Reviewer 2: ```suggestion
The `@commercetools-frontend/mc-http-server` package has been deprecated and will not receive any updates.
```

> With the usage of the `compile-html` command there is no need to have a pre-configured HTTP server anymore. If you are using this package, we recommend to use any other HTTP server package to serve your static files.

Reviewer 2: Maybe we can recommend one.

Author: You mean like express?

Reviewer 2: 😄 I don't mind. Express or serve or any package. We don't have to was just a thought.

<Info>

Remember that after building your production bundles you need to [compile the Custom Application](/deployment/compiling-a-custom-application#deployment) for production usage.

</Info>

## New command `serve`

> We added a new command `mc-scripts serve` that can be used to start your Custom Application **locally** in `production` mode, after having it compiled.

Reviewer 2: ```suggestion
We added a new command `mc-scripts serve` that can be used to start your Custom Application **locally** in `production` mode after it has been compiled.
```

The command starts an HTTP server to serve the static assets from the `public` folder.

<Warning>

> This command should only be used locally to test the Custom Application in production mode, as it contains the development routes for `/login` and `/logout`.

Reviewer 2: ```suggestion
This command should only be used locally to test the Custom Application in production mode, as it contains the development routes for `/login` and `/logout`. Do not use this command to serve your Custom Application in production.
```

</Warning>

Before:

```json title="package.json"
{
  "start:prod:local": "NODE_ENV=production MC_APP_ENV=development dotenv -- mc-http-server --use-local-assets"
}
```

After:

```json title="package.json"
{
  "compile-html:local": "NODE_ENV=production MC_APP_ENV=development dotenv -- mc-scripts compile-html --transformer @commercetools-frontend/mc-dev-authentication/transformer-local.js",
  "start:prod:local": "yarn compile-html:local && mc-scripts serve",
}
```

## Removing the `extract-intl` command

The `mc-scripts extract-intl` command has been removed in favor of the [official `@formatjs/cli` package](https://formatjs.io/docs/tooling/cli#extraction).

> We recommend to update your script to extract Intl messages to use the `formatjs extract` command.

Reviewer 2: ```suggestion
We recommend to update your script to extract intl messages to use the `formatjs extract` command.
```

Author: In our docs, we always write it with capital letter, as it refers to the Intl object/feature.

Before:

```json title="package.json"
{
  "i18n:build": "mc-scripts extract-intl --output-path=$(pwd)/src/i18n/data 'src/**/!(*.spec).js' --build-translations"
}
```

After:

```json title="package.json"
{
  "extract-intl": "formatjs extract --format=$(pwd)/intl-formatter.js --out-file=$(pwd)/src/i18n/data/core.json 'src/**/!(*.spec).js'"
}
```

> where the `intl-formatter.js` is something like this (depending on your translation tool):

Reviewer 2: ```suggestion
where the `intl-formatter.js` should be defined as as something like this depending on your translation tool's needs:
```

```js title="intl-formatter.js"
// https://formatjs.io/docs/tooling/cli#extraction
exports.format = function format(extractedMessages) {
  return (
    Object.keys(extractedMessages)
      // transform strings to lowercase to imitate phraseapp sorting
      .sort((a, b) => a.toLowerCase().localeCompare(b.toLowerCase()))
      .reduce(
        (allMessages, messageId) => ({
          ...allMessages,
          // Return a simple key/value JSON object.
          [messageId]: extractedMessages[messageId].defaultMessage,
        }),
        {}
      )
  );
};

> ```

Reviewer 3: I don't think it is necessary to add this examples.
it would be enough and suffice to link to the message extraction of `formatjs`
https://formatjs.io/docs/getting-started/message-extraction

Author: This file is what we have also included in the starter app and the playground app. I think it wouldn't hurt to have it documented here as well.

Reviewer 3: very well.

# Changes about the `experimentalRenderAppWithRedux` from test utils

The experimental render method `experimentalRenderAppWithRedux` from the `test-utils` has been removed.

Instead, you should pass the `disableApolloMocks` option to the `renderApp` and `renderAppWithRedux` methods. When this option is set to `true`, the real `ApolloProvider` is rendered instead of Apollo's `MockProvider`.

This is useful if you want to mock requests at the network level, for example when using [Mock Service Worker](https://mswjs.io/).

> Additionally, you can also pass a custom `apolloClient` instance together with the `disableApolloMocks` option. This is only useful when your Custom Application uses a custom `apolloClient`, for example for configuring the cache policies.

Reviewer 2: ```suggestion
Additionally, you can also pass a custom `apolloClient` instance together with the `disableApolloMocks` option. This is only be needed when your Custom Application uses a custom `apolloClient`, for example for configuring the cache policies.
```
